### PR TITLE
added dlock (distributed lock) -alpha

### PR DESCRIPTION
**Pull request by ilgooz**
Hey, here is the initial version of dlock. Thanks!

closes https://github.com/lieut-data/mattermost-plugin-api/issues/20

### CODE REVIEW

File: dlock/dlock.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	// lockRefreshInterval used to determine how long to wait before refreshing
+	// a lock's expiry time.
+	lockRefreshInterval = time.Second
+
+	// lockTryInterval used to wait before trying to obtain the lock again.
+	lockTryInterval = time.Second
+)
+
+// API is a portion of pluginapi to keep locks' state and log errors.
+type API interface {
```
Comment (ilgooz) [inline dlock/dlock.go]: I think it's best to split this interface into two, as Logger and Store and accept them separately at New(). That way it's easier for caller to pass necessary implementations and to avoid possible name conflicts inside possible big variations of API interfaces. I suggest this as an idiomatic pattern for plugins.

This can be accomplished in a separate PR.

File: dlock/dlock.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-afterC:
+		}
+	}
+}
+
+// TryLock tries to obtain the lock immediately.
+// err is only filled on system failure.
+func (d *DLock) TryLock() (isLockObtained bool, err error) {
```
Comment (hanzei) [inline dlock/dlock.go]: 1/5 In general I don't see the necessity to name the return values. The go doc comment should make clear, what the method does.

File: dlock/dlock.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+	}
+	if isLockObtained {
+		d.startRefreshLoop()
+	}
+	return isLockObtained, nil
+}
+
+// startRefreshLoop refreshes an obtained lock to not get caught by lock's TTL.
+// TTL tends to hit and release the lock automatically when plugin terminates.
+func (d *DLock) startRefreshLoop() {
```
Comment (hanzei) [inline dlock/dlock.go]: Clever way to do this :+1:

File: dlock/dlock.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	ctx, cancel := context.WithCancel(context.Background())
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		t := time.NewTicker(lockRefreshInterval)
```
Comment (hanzei) [inline dlock/dlock.go]: Maybe try to get the lock a little bit before the TTL expires to avoid races?

File: dlock/dlock.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+	}()
+	d.refreshCancel = cancel
+	d.refreshWait = &wg
+}
+
+// Unlock unlocks Lock().
+// use Unlock() exactly like sync.Mutex.Unlock().
+func (d *DLock) Unlock() error {
+	d.refreshCancel()
+	d.refreshWait.Wait()
```
Comment (hanzei) [inline dlock/dlock.go]: Can you elaborate why we need wait here?

File: dlock/dlock_test.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
+package dlock
```
Comment (hanzei) [inline dlock/dlock_test.go]: Can this be in `dlock_test` instead?

File: dlock/dlock_test.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+func TestLock(t *testing.T) {
+	dl := New("a", dlocktest.New())
+	var wg sync.WaitGroup
+	wg.Add(3)
+	for i := 0; i < 3; i++ {
+		go func() {
+			defer wg.Done()
+			for i := 0; i < 10; i++ {
+				dl.Lock(context.Background())
```
Comment (hanzei) [inline dlock/dlock_test.go]: Can you please check all the returned errors?

File: dlock/dlock_test.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+	"github.com/lieut-data/mattermost-plugin-api/dlock/dlocktest"
+	"github.com/stretchr/testify/require"
+)
+
+// TODO(ilgooz): test all branches including related ones to Store errors and ExpireInSeconds.
+// TODO(ilgooz): can move tests from sync/mutex_test.go.
+
+func TestLock(t *testing.T) {
+	dl := New("a", dlocktest.New())
```
Comment (hanzei) [inline dlock/dlock_test.go]: Maybe it makes sense to also test dlock with a mocked API instead of you in-memory implementation to make sure the API calls work as expected.

File: dlock/dlock_test.go (hunk 7 above)
Comment (hanzei) [inline dlock/dlock_test.go]: Maybe alter the test to also check the rough order in which the locks are acquired.

**Comment by lieut-data** [comment]
@ilgooz / @hanzei, I'm going to go ahead and merge this as-is, allowing us to iterate on the alpha efforts here and eventually unblock our larger goal of promoting this repository to the `mattermost` organization.
